src/Components/Components/src/RenderTree/StackObjectPool.cs

```diff
@@ -0,0 +1,73 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.AspNetCore.Components.RenderTree
+{
+    // This is a very simple object pool that requires Get and Return calls to be
+    // balanced as in a stack. It retains up to 'maxPreservedItems' instances in
+    // memory, then for any further requests it supplies untracked instances.
```

Comment:
I assume no need to be thread-safe then?

Reply:
No, it's not making any claims of thread-safety (like most other collection structures). We only use it within the context of a single renderer, where we've already gone to great lengths to ensure calls and hence renders are serialized.